crates/noirc_driver/src/lib.rs

```diff
@@ -22,6 +22,19 @@ pub struct Driver {
     language: Language,
 }
 
+#[non_exhaustive]
+#[derive(Debug, Default)]
+pub struct CompileOptions {
```

Comment:
I don't see the point in having both this struct and `NargoCompileOptions` which is an exact duplicate with another name. Is the `#[non_exhaustive]` preventing us from removing this one? Why do we have it again?

Reply:
There's a couple of reasons for this:
- We can't derive `Args` on `CompileOptions` without making `noirc_driver` aware of clap.
- We're separating the interface of the CLI from that of the compiler so that we can modify one without affecting the other. Otherwise whenever we added a new field to `CompileOptions` it would immediately end up in the CLI (unless we explicitly hid it).

`#[non_exhaustive]` by itself doesn't affect this, I'm just anticipating that we'll add more compiler options in the future. If we provide sensible defaults for these then we can avoid having breaking changes between `noirc_driver` and `nargo`.

Reply:
> We're separating the interface of the CLI from that of the compiler so that we can modify one without affecting the other. Otherwise whenever we added a new field to CompileOptions it would immediately end up in the CLI (unless we explicitly hid it).

Isn't that the entire point of the struct, to provide the CLI options? Perhaps we can rename it `CliOptions` to be even more explicit and remove the duplication.

Reply:
Partially, `NargoCompileOption` exists so that we can define CLI options related to compilation in a single place rather than needing to add a new option to every relevant command.

`CompileOptions` exists to have a compact way to pass options into the compiler without having N boolean flags on each entrypoint.  Not all of these options must necessarily come from the user through the CLI, some options may be injected based on config files, etc.

Renaming `NargoCompileOption` to `CliCompileOptions` may be clearer however (as it's specific to options which are set in the CLI).

Reply:
I think it is fine to pass around the CliCompileOptions directly to avoid the duplication. If we ever do add more options in other places like config files we can just add a wrapper struct to pass around that contains CliCompileOptions internally rather than duplicating it.

Reply:
I disagree on this. If we want to use this struct in the CLI and in `noirc_driver`,  our alternatives are to make `noirc_driver` aware of either `clap` or `nargo`. Neither is a good option as either we get a unwanted dependency in `noirc_driver` just so `nargo` can use it or a cyclic dependency.

Reply:
If we see noirc_driver as an analogue for rustc the same way nargo is an analogue for cargo then I think it is fine for it to be aware of clap. Currently we provide no exact rustc alternative and all projects must be conform to nargo's structure but I could see a reasonable argument for exposing a noirc via noirc_driver in the future to get around this.

Reply:
I'd prefer to keep `noirc_driver` as a "pure" library until we have a `noirc` executable similar to `rustc` and adding knowledge of CLI arguments would violate this.

I'd accept some duplication to avoid making `noirc_driver` aware of CLIs but it wouldn't be an earthshattering issue if you feel strongly about this.

Reply:
I've removed the separation between `CompilerOptions` and `NargoCompilerOptions`

Reply:
Ah, one sec. I need to remove some `from`s